catshadow/client.go

```diff
@@ -1378,30 +1385,29 @@ func (c *Client) Online() error {
 // goOnline is called by worker routine when a goOnline is received. currently only a single session is supported.
 func (c *Client) goOnline() error {
 	c.connMutex.RLock()
-	if c.online || c.connecting || c.session != nil {
+	if c.connected == StateOnline || c.connected == StateConnecting {
 		c.connMutex.RUnlock()
 		return errors.New("Already Connected")
 	}
 	c.connMutex.RUnlock()
 
```

Comment:
There's a race here where after `RUnlock()` and before `c.connMutex.Lock()` the state could change, meaning two `goOnline()` can be running at the same time, launching two parallel `c.client.NewTOFUSession()`s.
Since `sync` doesn't seem to have an *upgradeable lock* (???) we could roll our own by having an exclusive mutex guard lock acquisition for the `sync.RWLock` and chuck that in our `utils` library.
In this specific case though, I think it seems better to just take the write Lock up front:
```go
c.connMutex.Lock() // could also be a TryLock where we return early if it fails?
switch c.connected {
    case StateOnline: {
       c.connMutex.Unlock()
       return errors.New("Already Online")
    }
    case StateConnecting: {
       c.connMutex.Unlock()
       return errors.New("Already Connecting")
    }
}
c.connected = StateConnecting
c.connMutex.Unlock()
```
Alternatively I guess we could explore using atomics for state transitions?

Reply:
see d44d376bf51c93c49f51eed1bab622df7599b269

Reply:
(thanks)

Reply:
https://github.com/katzenpost/katzen/pull/29#issuecomment-1595825465
new code takes a read-only lock `RLock()` then tries to relinquish a read-write lock (`Unlock()`), this results in a crash

Reply:
> [katzenpost/katzen#29 (comment)](https://github.com/katzenpost/katzen/pull/29#issuecomment-1595825465) new code takes a read-only lock `RLock()` then tries to relinquish a read-write lock (`Unlock()`), this results in a crash

fixed in c3a210d0ae74ec152cf6807b45aa50ee56d74b15